compiler/src/dotty/tools/dotc/transform/ArrayApply.scala

```diff
@@ -0,0 +1,53 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import MegaPhase._
+import Contexts.Context
+import Symbols._
+import Types._
+import StdNames._
+import ast.Trees._
+import dotty.tools.dotc.ast.tpd
+
+
+/** This phase rewrites calls to `Array.apply` to primitive array instantion.
+ *
+ *  Transforms `scala.Array.apply([....])` and `scala.Array.apply(..., [....])` into `[...]`
+ */
+class ArrayApply extends MiniPhase {
+  import tpd._
+
+  override def phaseName: String = "arrayApply"
+
+  override def transformApply(tree: tpd.Apply)(implicit ctx: Context): tpd.Tree = {
+    if (tree.symbol.name == nme.apply && tree.symbol.owner == defn.ArrayModule) { // Is `Array.apply`
+      tree.args match {
+        case CleanTree(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: ct :: Nil
+            if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) && elideClassTag(ct) =>
+          seqLit
+
+        case elem0 :: CleanTree(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: Nil
+            if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) =>
+          tpd.JavaSeqLiteral(elem0 :: seqLit.elems, seqLit.elemtpt)
+
+        case _ =>
+          tree
+      }
+
+    } else tree
+  }
+
+  // Only optimize when classtag is `ClassTag.apply` or `ClassTag.{Byte, Boolean, ...}`
+  private def elideClassTag(ct: Tree)(implicit ctx: Context): Boolean = {
+    ct.symbol.maybeOwner.companionModule == defn.ClassTagModule
+  }
+
+  object CleanTree {
+    def unapply(tree: Tree)(implicit ctx: Context): Some[Tree] = tree match {
```

Comment:
Is this necessary for the common case where the user doesn't manually add type ascriptions or blocks himself ? If it's not necessary then I'd say we shouldn't bother trying to optimize these edge cases.

Reply:
Those are to remove artefacts of inlining. They will be useful if we want to optimize the instantiation an opaque type containing an array (such as in `IArray`).

Reply:
I see, but it'd be even nicer if inlining didn't produce such artfacts :)

Reply:
Only the `Typed` is an artefact. But now I see it is an artefact of unpacking a varargs into another one. In Scala2 there is an equivalent [cast striped in the same place](https://github.com/scala/scala/blob/d030172d7ef807d85391d32c5456b1b97b15a402/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L540).